### PR TITLE
Improve Documentation Style

### PR DESCRIPTION
**Pull request by marianoguerra**
Improvements in documentation style, no major layout changes, here's a blog post describing the changes with screenshots:

http://marianoguerra.org/posts/improving-official-erlang-documentation.html

here is the new documentation online: https://marianoguerra.github.io/otp/doc/

### CODE REVIEW

**Comment by vladdu** [comment]
Excellent! There will probably be opinions on some choices (for example, is a line width of 1.5em too large?), but the important thing is that there is a good and almost complete starting point. The only thing I see missing is having the layout responsive, so that it is usable even on narrow displays (phones).

**Comment by dgud** [comment]
The size of the font in the word "Types:" is too large IMO, for example in:
[lists:foldl/3](https://marianoguerra.github.io/otp/lib/stdlib-3.1/doc/html/lists.html#foldl-3)

**Comment by marianoguerra** [comment]
@vladdu I plan to work on a responsive layout later, didn't wanted to make this PR to disruptive.

@dgud yep, seems large, I was also thinking on removing the colon, since it's already a title.

**Comment by sverker** [comment]
Links with `<c>` markers look the same as regular `<c>text</c>`.

**Comment by psyeugenic** [comment]
I think this is definitely a step in the right direction. Work on the HTML layout will be worked on and perhaps completely redone, so we shouldn't spend an abundant of energy fixing the old doc. Now, there's no time-frame for the new docs so your work here will be a good mid-step I think.

- I also agree with @sverker , I would like to see the links in `<c>` tags. 
- I also think the font is pretty large in general but especially for types.
- It would be great to have syntax highlighting for type-specifications (is that easily realizable?)
- I would like to have the width for the content bounded somehow and not flow across the whole screen (more readable?)

The last two items is not something we need to fix in this PR IMHO.

My 2 cents.

**Comment by marianoguerra** [comment]
@sverker could you point to an example of a <c> link so I know what those are? :)

@psyeugenic I've noticed that in chrome the font is quite large, I don't specify a size AFAIK so it seems to pick up some default, will look into it.

will take a look at the syntax highlight for types, right now they generate some markup which I would have to remove to get the syntax highlighter to work.

I added a limit of 80 characters to paragraphs so they won't continue to the right if there's more space (which makes it hard to follow lines), it's here: https://github.com/marianoguerra/otp/blob/improve-docs-style/lib/erl_docgen/priv/css/otp_doc.css#L15 do you see it happening in my version? if so, can you link somewhere where it happens?

**Comment by sverker** [comment]
https://marianoguerra.github.io/otp/erts-8.1/doc/html/erl_nif.html#enif_alloc_binary

`size` and `bin` are text while `enif_release_binary` is a link.

**Comment by marianoguerra** [comment]
@sverker done

**Comment by marianoguerra** [comment]
just checked, and I'm not setting a font-size, chrome seems to default to the recommended size which is 16 css pixels (scale dependent on screen), firefox in my machine displays text smaller, I can set it to 14px which is below what it's recommended and we can see how it displays on different displays, OSs and configs.

https://developers.google.com/speed/docs/insights/UseLegibleFontSizes

there it says that default line-heigt should be 1.2em so I set it to that.

**Comment by psyeugenic** [comment]
I fiddled with `p { max-width: 80em }` .. `60em` looks nicer for me but I'm eerily close to bike-shedding now though. All items, i.e. `<li>`, will not respect `max-width` in this case, right?

I will rest until others give their input as well.

**Comment by marianoguerra** [comment]
@psyeugenic I can add it to li's and any other element too, just thinking where they may go that long :)

**Comment by psyeugenic** [comment]
Well, the top page, `index.html` has items that run fairly long. =)

**Comment by vladdu** [comment]
@marianoguerra If you set "max-width: 80em" (or 70em) on the #content div, then all content will be constrained to that value.

**Comment by marianoguerra** [comment]
@vladdu done

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by essen** [comment]
The font has become too small now. Please try to keep it around the same size as the original Erlang docs, otherwise it's hard to read on small HiDPI screens.

**Comment by marianoguerra** [comment]
@essen there wasn't a font size set before, you can see it here: https://github.com/erlang/otp/pull/1215/files#diff-976da1697145f78558665c67351b4a3aR13 (search for font-size), I think you were seeing your browser's default size, I can set it to 16px as recommended in the document I linked before, that should be good for everybody but some may see it "too big".

**Comment by essen** [comment]
Too big is better than too small IMO. Better see than not.

**Comment by richcarl** [comment]
Not sure what max-width you're discussing, but if it's for the general descriptive text, then I'm not seeing it (in your docs on github). I'm getting lines of 160+ characters, and it's close to unreadable. (Firefox on Mac, retina screen.)

**Comment by marianoguerra** [comment]
@essen made font 16px as recommended, if it's to big users can "zoom out" with ctrl -
@richcarl I reduced max-width to 60em can you notice a difference? I don't have a machine which retina or similar to test :/

**Comment by essen** [comment]
Thanks, that's better. On the other hand the monospace text is probably too big. By that I mean that on my screen at least, the little grey box around the monospace text can end up eating up lines above or below. If there is monospace text right above with underscores, I do not see the underscores.

I can see this in the "Read and write Erlang terms" part in erl_nif, for example.

**Comment by richcarl** [comment]
Now the font size is nicer, but I still see no effect of max width. This is what a typical page looks like in full screen view:
screen shot 2016-10-25 at 15 31 28

**Comment by vladdu** [comment]
@richcarl This is the page at the specified max width. Maybe your screen width is exactly matching it? If you pres Ctrl-- a couple of times, you should be able to see it zoomed out and it shouldn't take the whole width.

**Comment by richcarl** [comment]
I did an experiment with the css editor in firefox, and it turns out that it's the em unit that is causing the problem. I guess it's a question of what font the ems are taken to refer to. In practice, I get something like 28em -> 65 characters. 60em is large enough to be off-screen.

**Comment by vladdu** [comment]
@richcarl In the css editor, select the body and add "font-size: 1em;", does it help? Try also "font-size: 100%;"

Many sites use a "normalized css" that removes the differences between the defaults of each browser. I suppose it would help in this case too. The reference one is at https://necolas.github.io/normalize.css/. Other sites use a reset css (which removes all default styling), but then the styling has to be added back.

**Comment by dgud** [comment]
Missing line breaks?

[add_report_handler(Handler) -> any()add_report_handler(Handler, Args) -> Result](https://marianoguerra.github.io/otp/lib/kernel-5.1/doc/html/error_logger.html#add_report_handler-1)

**Comment by vladdu** [comment]
re @essen comment: the overlapping boxes can be fixed with `body{line-height:1.3em;}` and `span.code{padding:0;}`, which seems ok but may have corner cases too.

**Comment by richcarl** [comment]
I think that for specifying width, the ex unit is more useful. To quote w3.org: "The ex unit is rarely used. Its purpose is to express sizes that must be related to the x-height of a font. The x-height is, roughly, the height of lowercase letters such as a, c, m, or o. Fonts that have the same size (and thus the same em) may vary wildly in the size of their lowercase letters, and when it is important that some image, e.g., matches the x-height, the ex unit is available."

The em unit is equivalent to the height of the font. For a tall narrow font, that means that you get an awful lot of characters in 80em. However, setting max-width: 80ex in #content gives a layout with approximately 80 characters per line. (I tried removing the declaration for p and it seems to work fine without it.)

**Comment by vladdu** [comment]
@richcarl You get different results with different fonts. I would rather use 60em, or 50em, or whatever fits, because this is what is used generally and it's easier to relate to the values. (I.e. we introduce an additional ratio that depends on the font, em/ex)

**Comment by richcarl** [comment]
Ideally, the ex unit would be better than em, but apparently, some browsers just define 1ex = 0.5em, so there's not much point in it (but hardly wrong either). But getting different results with different fonts would exactly be the point. :-) Anyway, choosing 42em gives me a reasonable layout with about 80 chars per line. With 60em, we're talking 120 chars per line, which is much too much.

**Comment by Jianru-Lin** [comment]
Nice work! I like it!

**Comment by marianoguerra** [comment]
@dgud should be fixed, nice catch!

also Types title should be a little smaller

**Comment by richcarl** [comment]
This is the layout with #content max-width 42em (window in fullscreen mode):
screen shot 2016-10-25 at 20 08 50

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by unbalancedparentheses** [comment]
Awesome. I think it is a great way to start. I will try to add a few lines of CSS to make typography and spacing even better, based on recommendations of the [Practical Typography book](http://practicaltypography.com/).

**Comment by richcarl** [comment]
@psyeugenic I'm curious: when you said that "60em looks nicer to me", what kind of character count per line are you getting with that number? Screenshot?

**Comment by psyeugenic** [comment]
@richcarl I get about 120 - 130 characters with 60em too. I'm not totally bounded by 80 char limit on docs or on my terminals. But I do want some boundaries and I simply stated that 60em was "nicer" than 80em for me.

This is how it looks for me with 60em:
screenshot-new-doc

**Comment by richcarl** [comment]
Well, I wasn't trying to emulate the 80 chars of a terminal. But the typographical recommendations for prose is between 50 and 70 (try picking up any book from your shelf and count if you want to verify). For technical texts, it seems that a bit longer is better, because you don't read it in quite the same way and it gets more important to squeeze in more info vertically, but I'd say that 90 is the limit. When lines get too long, your eyes need to move too much, first taking in the left part of the line, then the right part of the line, and then when you go back left again you often lose track of which line is the next. It screws up the flow lf reading.

Setting max-width to 42 (just some arbitrary number, don't know why I like it...) gives around 85 characters per line in a normal font, so I vote for that. (See my screenshot above.)

**Comment by richcarl** [comment]
@marianoguerra While you're at it, how about my request to set max-width to 42 (or 40 if you prefer)?

**Comment by marianoguerra** [comment]
@richcarl done

I also improved image position (now works like tables), title line-height for titles that have more than one line and do's and dont's boxes now look like notes and warnings.

please check on your browsers, here's a page to try: https://marianoguerra.github.io/otp//erts-8.1/doc/html/erl_dist_protocol.html (clean your browser's cache if you visited it before).

question: wouldn't it be better if text alignment was justify instead of left?

**Comment by richcarl** [comment]
Thanks, that's much better, I think.
Table 12.6 on your example page gets pretty wide, and I think it's because the fixed-width font used in the table is the same size as in the running text, but in tables the font size (for all fonts) should probably be a little smaller.
Text alignment: experiment with justified and see what you think. A problem is that technical texts tend to contain many long words, code snippets and identifiers that cannot be nicely broken over lines, so a jagged right can be a better choice than justified text with a lot of ugly extra white space between words just to fill out lines. It depends on the text.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by lthor** [comment]
We will merge this into OTP after the 19.2 release.

**Comment by richcarl** [comment]
I think it would be good to try to fix the table issue that I mentioned first. The tables get way too large compared to the main text; see the example page linked to above.

**Comment by psyeugenic** [comment]
I would also like to see some squashing of commits.

**Comment by marianoguerra** [comment]
sorry for the delay, I will work on this this week or the next

**Comment by lthor** [comment]
Hi,
can you also take a look at the examples. It's the same problem as with the tables that they get too large and go beyond the borders. I know it's because we don't break the lines for them but it will probably be better if you make the font size a bit smaller as with the tables.

**Comment by marianoguerra** [comment]
@lthor code samples have a smaller text size
@richcarl same for tables, please check them here: https://marianoguerra.github.io/otp//erts-8.1/doc/html/erl_dist_protocol.html (clear cache just in case)

@psyeugenic how would you like me to squash them, all in one commit? split them in a couple of them? how many and how do I group them?

**Comment by marianoguerra** [comment]
one question, should I make function types and type declarations text smaller too? check here: https://marianoguerra.github.io/otp/lib/sasl-3.0.1/doc/html/release_handler.html#install_release-1

**Comment by marianoguerra** [comment]
one question, should I make function types and type declarations text smaller too? check here: https://marianoguerra.github.io/otp/lib/sasl-3.0.1/doc/html/release_handler.html#install_release-1

**Comment by richcarl** [comment]
@marianoguerra That's better, but perhaps it ended up just a bit too small this time. Changing it to 75% looks better (possibly even 80%, but then some tables hit the margins; on the other hand, some of those tables are just unreasonably wide anyway) - both for examples and tables. It looks like you've mostly used % for other relative font sizes so I suggest you do that uniformly unless there is a specific reason for using em. (I see that main-description also uses em.)

**Comment by richcarl** [comment]
Alternatively, follow the advice of the addendum in this post: http://kyleschaeffer.com/development/css-font-size-em-vs-px-vs-pt-vs/

**Comment by essen** [comment]
Please don't make the text this small. The code snippets are hard to read now. The current "normal" text size is really the smallest I can take.

**Comment by richcarl** [comment]
@marianoguerra Yes, I think it could be worth trying to reduce the size of those type declarations. Maybe to 90% or so.

**Comment by marianoguerra** [comment]
@richcarl 0.8em is the same as 80% it's 0.8 times the size of the default font size which is 80% of 1em AFAIK. check this example: https://codepen.io/anon/pen/qRmrbm screenshot attached to see how it looks on my machine
shot-2017-01-22_00-16-34

@essen I would prefer not to shrink them, maybe mark wide tables and examples with an extra class to shrink only those or just arrange them on some other way?

**Comment by marianoguerra** [comment]
@unbalancedparentheses can you bring your typography-foo to the table? :)

**Comment by essen** [comment]
I understand that smaller columns for making text readable is good, but it's not so good for code. I've seen books with code snippets spanning over the margins, perhaps this is what's needed here. Allow them to have the same width as the current docs.

**Comment by richcarl** [comment]
@marianoguerra Yes, by definition 1em should be equivalent to 100% - they should both be interpreted as relative to the parent element font size. Historically though, it seems that some browsers (of course) used to interpret these a bit differently, so that em didn't always scale properly when people zoomed in. Probably not much of an issue today. Anyway, I'm just saying that whatever you pick, it should probably be done consistently throughout the css. Right now there's a mix of % and em for font-size.

**Comment by marianoguerra** [comment]
@richcarl unified some paddings that were in pixels and all font-sizes that where in % now are in ems.

I think all the feedback was addressed, if so, I will squash the commits, what's the preferred way, all commits in one big commit? should I create a new branch with that commit and open a new PR?

**Comment by marianoguerra** [comment]
I rebased the changes on top of master as a single commit here: https://github.com/marianoguerra/otp/commits/improve-docs-style-1 is that ok? should I open a PR with that?

**Comment by vans163** [comment]
Going to give my 2 cents here, please take it with a grain of salt.

I kind of like the original style of the documentation, there should be an option to switch back.  

The problem in the erlang docs is sometimes there is lack of key details that cannot be found out unless you peek at the actual source code or do trial and error.

The text sizes need adjusting across the board. Headers are too big. 
The text sizes in the code markdown blocks seem a little too small maybe the font.
Ex: look at the NOTE https://marianoguerra.github.io/otp/erts-8.1/doc/html/escript.html,
its really hard to read due to the tiny size. 
" Summary of Options" href there seems to be using a totally different font from the rest of the text.

The grey highlight color need tuning/styling, hurts my eyes.

The white folders on white background seem to blend in, I cannot quickly determine if its a folder or part of the white.

The restyled white file is too white, its hard to tell what it is again like the folders. 

I also think a useful enhancement to the docs would be to have autogenerated xrefs like Doxygen, so you can quickly jump around and see what is called by what without peeking at OTP sources each time.

Great initiative and I hope it keeps going!

**Comment by essen** [comment]
Thanks for +1ing on the code snippets font being too small. I simply can't read them.

**Comment by psyeugenic** [comment]
I think we should merge these changes to master asap.

I do agree that more tweaks are in order. As for font size on code snippets and function declarations - the contrast is certainly bigger between them now. I'm fine with the font size for code snippets, however I think the style for function declaration is too clunky. It's too big, too fat or too messy. Especially when there are a lot of types and input specifications to the function.

Regardless of that, I think we should merge the current changes and then do more tweaks after the merge.
@marianoguerra yes, open a PR with that commit.

**Comment by essen** [comment]
Allow me to insist again on the font size in code snippets. It is *not readable*. This is not a figure of speech. The documentation is simply not usable in the current state. To make snippets readable, I have to go from my normal zoom of 125% to 175% (!!), at which point everything else is so big I have very little content on my screen anymore. This should be fixed before any merge happens, otherwise it will be forgotten and result in many users ending up with an unusable documentation, whereas the current one was perfectly fine as far as readability is concerned.

If it looks fine on your big screen, try on your phone. You will zoom on the snippets.

**Comment by psyeugenic** [comment]
Funny. For me it's absolutely perfect on my phone. Crisp and clear. The menu to the left is a bit big for the phone and the function declarations are annoyingly big but otherwise perfect.

Anyways, since several of you obviously have a hard time reading that size we should probably change before committing. Just gently increase it a minimal step.

**Comment by marianoguerra** [comment]
hi, if there's no problem and given that there seems to be some consensus I would like to do two things before submitting the new PR:

* increase the code sample size to where it was
* decrease the function type declaration size

all the other request may be discussed in the new PR or open as issues mentioning me (or I can open them) and we can discuss them individually, see if there's consensus and open a PR for each after this one is merged, as it is now some feedback involves changing the markup generated which in this PR I'm not changing.

The idea of this PR was to get the ball rolling and try to get people involved so we can make this an ongoing effort and not a single PR.

@essen just to have a clear picture of what you are referring, I already agree that code size is too small, but just to be sure that there's no other problem like browser, OS or settings involved, can you attach a screenshot of one of the code sections you refer to?

**Comment by essen** [comment]
In my usual browser, Opera, at 125% (my default, and corresponds roughly to the DPI on the rest of the system):

docs_opera_125percent

The only place where I increase the zoom further is a Web comic. Github has smaller code size than normal text, but the difference is minimal and it is perfectly readable, for example.

For reference, in Chromium, at 100%:

docs_chromium_100percent

**Comment by vladdu** [comment]
as @vans163 pointed out, the code is even smaller when used in a note, because the size is relative to the enclosing container and the note has already smaller text. 

The proper solution would be to use rem units instead of em, which always use the root element's size as base (see https://www.sitepoint.com/understanding-and-using-rem-units-in-css/). All current versions of browsers support these units (see http://caniuse.com/#feat=rem), only IE has some issues with rems in versions 9 and 10 (and no support for earlier). This might be a problem in corporate environments where these older versions are still used...

**Comment by lthor** [comment]
Hi,
we have now tested this pull-request in our nightly builds and found two problems.

The first is the link from the logo to the top of the documentation. The documentation is not always accessed via a web server and therefore we have a problem with a link to the directory. It has to be directly to the index.html file instead.

The second is the text under the title on the top index file. The sentence "Welcome to Erlang/OTP, ....." is formatted to two lines where the second line is so close that the first line partially is overwritten.

**Comment by marianoguerra** [comment]
@IngelaAndin fixed both and published the changes here: https://marianoguerra.github.io/otp/doc/ you can try the logo link here: https://marianoguerra.github.io/otp/doc/reference_manual/users_guide.html

**Comment by vladdu** [comment]
The docs for several groups of apps  (basic, database, etc)  are missing from  https://marianoguerra.github.io/otp/doc/ . Is this intentional?

**Comment by marianoguerra** [comment]
@vladdu I restored some, the only missing I know of is the system folder, for some reason it wasn't generated in my last build, will check it.

Note: the gh-pages branch in my fork of otp is a little hack I did to show the current status, so I "merge" the generated docs semi manually, forgot to add some folders,

**Comment by lthor** [comment]
Hi Mariano,
we don't want any merge commits on the topic branches. So can you remove the merge commit from your branch by resetting it to the "improve line-height ..." commit (7c32092). And then do a rebase to master instead of the merge.

BR Lars

**Comment by marianoguerra** [comment]
resetted and rebased, should I squash them into one now that I'm rewriting history? :)

**Comment by marianoguerra** [comment]
here's a branch with a squashed commit: https://github.com/marianoguerra/otp/tree/improve-docs

we can keep this for the discussion, close it and I can open a new PR with the new branch

**Comment by marco-m** [comment]
Any news?

**Comment by lthor** [comment]
It is scheduled for OTP 20 and is running in our internal tests.


/Lars

OTP Team


________________________________

**Comment by lthor** [comment]
Thanks for the contribution, merged to master branch.
